forem/forem
# Try fixing comment like count bugs

## What type of PR is this? (check all applicable)


## Related Tickets & Documents

* Closes :bug: https://github.com/forem/forem/issues/18312
* Closes :bug: https://github.com/forem/forem/issues/18338

## Description

I wrote this up in https://github.com/forem/forem/issues/18312#issuecomment-1241611473, I'll just include the relevant bits:

* Prior to https://github.com/forem/forem/pull/11283, we had special treatment for a newly created comment that would assume it had 1 like, by the current user.
* Now that this starts at zero, the first time a user interacts with a like (manually self-liking), because like-ing is toggle-ing, the initial like would be removed, thus the count is deprecated in JS, and we might see a [negative like](https://github.com/forem/forem/issues/18338).
* However, because we are creating the initial self-like in a background job, if we restore the UI assumptions (https://github.com/forem/forem/pull/18354 for example), there's still a race condition, where if the user interacts with the like before the background task finishes, if there is not _yet_ an initial like, we will often wind up creating **2 likes** for the user.
* @fdocr is [also correct](https://github.com/forem/forem/issues/18312#issuecomment-1231864379), we have a separate javascript that "polls" to update all comments for [reaction count updates](https://github.com/forem/forem/blob/main/app/assets/javascripts/initializers/initializeCommentsPage.js.erb#L17-L200). This then introduces a second race condition: if the poll polls before the background initial like finishes creating, it can reset the initial 1 like to 0. (It's not technically polling, it's triggered by the user, but the trigger is subtle and indirect and "feels like" background polling. There also seems to be a bit of a buggy inconsistency between the way the polling JS and the post-Ajax JS render 0 likes, but compared to everything else, this is very minor.)
* We do have a `public_reactions_count` on `comment` that _does_ cache the count -- however, this isn't used in the post-Ajax handler at all, only in the polling JS, so it's a factor in the various race-conditions, but I think this is more about the race between the various background behaviours more than the fact that we cache the count, per se.

I suspect we often see this more as a "local environment" issue because the timing of front-end and back-end background work in the dev environment is more tightly coupled (it's all happening on the same CPU) and not especially optimized.

From my perspective, it seems like the probable solution here would be to actually create the initial like in the same request where we create the comment. That would have us swimming a bit upstream (typically we want to move more things to background work), but given the way these overlap asynchronously, I think it's the only way to have the UI work consistently the way we want it to work. 


## QA Instructions, Screenshots, Recordings

1. Make a comment. It should have 1 like and be liked by you.
2. You can click the like and it will become unliked.
3. The like count/appearance should remain the same even if you refresh the whole page.

## Added/updated tests?

- [x] I ~~should actually write more~~ wrote some

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

      [Storybook](https://storybook.forem.com/) (for Crayons components)
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
      or in a [forem.dev](http://forem.dev) post
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media2.giphy.com/media/14wPFMQBn66xwI/giphy.gif?cid=ecf05e474zwy8ye3ehe07gwdrhazkpgh5zpaa8jsb44mxtna&rid=giphy.gif&ct=g)
